### PR TITLE
feat: insights orgs [CAL-1888]

### LINKED ISSUE #9360: [CAL-1874] Insights filters
[Orgs → Insights - 5. Organizations (Figma)](https://www.figma.com/file/mWENDNaJPdRfZMtA75MghZ/5.-Organizations?type=design&node-id=1378%3A74723&t=Avap0pYxqjof0wK5-1)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-1874](https://linear.app/calcom/issue/CAL-1874/insights-filters)</sub>

### PR DESCRIPTION
**Pull request by alannnc**
## What does this PR do?

- Handles insights if user logged in is owner/admin inside team org.

Fixes # (issue)
Fixes https://github.com/calcom/cal.com/issues/9360

https://www.loom.com/share/d149cc7b8dfd436487f404d472c1840c

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- [x] Start with a new fresh db.
- [x] Setup an org from /settings/organizations/new.
- [x] In your terminal navigate to your packages/prisma folder and run command yarn seed-insights.
- [x] Modify newly created insights team to be child of team org.
- [x] Navigate to your insights page as org owner/admin and use all the features.

## Mandatory Tasks
- [X] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/features/insights/components/AverageEventDurationChart.tsx [hunk 1]
```diff
@@ -12,7 +12,7 @@ import { LoadingInsight } from "./LoadingInsights";
 export const AverageEventDurationChart = () => {
   const { t } = useLocale();
   const { filter } = useFilterContext();
-  const { dateRange, selectedMemberUserId } = filter;
+  const { dateRange, selectedMemberUserId, isOrg } = filter;
```
Comment (alannnc) [inline packages/features/insights/components/AverageEventDurationChart.tsx:15]: new prop in UI that tell us if Team option selected is the team that is parent of other teams.

File: packages/features/insights/context/FiltersProvider.tsx [hunk 2]
```diff
@@ -69,13 +69,20 @@ export function FiltersProvider({ children }: { children: React.ReactNode }) {
...
   const [dateRange, setDateRange] = useState<FilterContextType["filter"]["dateRange"]>([
     startTimeParsed ? dayjs(startTimeParsed) : dayjs().subtract(1, "month"),
     endTimeParsed ? dayjs(endTimeParsed) : dayjs(),
     "t",
   ]);
+  const [initialConfig, setInitialConfig] = useState<FilterContextType["filter"]["initialConfig"]>({
+    userId: null,
+    teamId: null,
+    isOrg: false,
+  });
```
Comment (alannnc) [inline packages/features/insights/context/FiltersProvider.tsx:85]: Added this so I can save initialConfig and reinstate it when using clear filters button.

File: packages/features/insights/context/FiltersProvider.tsx [hunk 3]
```diff
@@ -89,18 +96,22 @@ export function FiltersProvider({ children }: { children: React.ReactNode }) {
...
           selectedFilter,
+          isOrg,
+          initialConfig,
         },
         setSelectedFilter: (filter) => {
           setSelectedFilter(filter);
           const userId =
             filter?.[0] === "user" ? selectedMemberUserId : selectedUserId ? selectedUserId : undefined;
           const eventTypeId = filter?.[0] === "event-type" ? selectedEventTypeId : undefined;
+          const memberUserId = filter?.[0] === "user" ? selectedMemberUserId : undefined;
```
Comment (alannnc) [inline packages/features/insights/context/FiltersProvider.tsx:107]: Was missing some clean up here when running clean filters action

File: packages/features/insights/server/trpc-router.ts [hunk 4]
```diff
@@ -12,10 +12,9 @@ import { EventsInsights } from "./events";
 
 const UserBelongsToTeamInput = z.object({
   teamId: z.coerce.number().optional().nullable(),
+  isOrg: z.boolean().optional(),
```
Comment (alannnc) [inline packages/features/insights/server/trpc-router.ts:15]: Most important changes where in this file and github just hides it

File: packages/features/insights/server/trpc-router.ts [hunk 5]
```diff
@@ -37,6 +37,47 @@ const userBelongsToTeamProcedure = authedProcedure.use(async ({ ctx, next, rawIn
...
+        accepted: true,
+      },
+    });
+    if (isOwnerAdminOfOrg && childTeam) {
+      return next({
+        ctx: {
+          ...ctx,
+          user: {
+            ...ctx.user,
+            isOwnerAdminOfParentTeam: true,
```
Comment (alannnc) [inline packages/features/insights/server/trpc-router.ts:74]: If no direct membership is found, we then look for parent teams and look if user session is admin/owner.

Comment (alannnc) [reply packages/features/insights/server/trpc-router.ts:74]: Then we pass down ctx a new prop that user is validated and if isOrg is passed in input we should allow user get child teams data.

File: packages/features/insights/server/trpc-router.ts [hunk 6]
```diff
@@ -107,10 +158,56 @@ export const insightsRouter = router({
...
-      if (teamId) {
+      if (isOrg && ctx.user.isOwnerAdminOfParentTeam) {
+        const teamsFromOrg = await ctx.prisma.team.findMany({
+          where: {
+            parentId: ctx.user.organizationId,
+          },
+          select: {
+            id: true,
+          },
+        });
```
Comment (alannnc) [inline packages/features/insights/server/trpc-router.ts:169]: Getting teams from org

File: packages/features/insights/server/trpc-router.ts [hunk 7]
```diff
@@ -107,10 +158,56 @@ export const insightsRouter = router({
...
+        if (teamsFromOrg.length === 0) {
+          return emptyResponseEventsByStatus;
+        }
+        teamConditional = {
+          id: {
+            in: teamsFromOrg.map((t) => t.id),
+          },
+        };
+        const usersFromOrg = await ctx.prisma.membership.findMany({
```
Comment (alannnc) [inline packages/features/insights/server/trpc-router.ts:178]: Then we convert those teams to users ids

File: packages/features/insights/server/trpc-router.ts [hunk 8]
```diff
@@ -965,40 +1370,41 @@ export const insightsRouter = router({
...
             },
-            where: eventTypeWhereConditional,
-          });
-          break;
+          },
+        },
+        where: eventTypeWhereConditional,
+      });
 
-        default:
```
Comment (alannnc) [inline packages/features/insights/server/trpc-router.ts:990]: Changed from switch to if conditional

File: packages/features/insights/server/trpc-router.ts [hunk 9]
```diff
@@ -244,10 +343,51 @@ export const insightsRouter = router({
 
       let whereConditional: Prisma.BookingTimeStatusWhereInput = {};
 
-      if (teamId) {
+      if (isOrg && ctx.user.isOwnerAdminOfParentTeam) {
+        const teamsFromOrg = await ctx.prisma.team.findMany({
```
Comment (alannnc) [inline packages/features/insights/server/trpc-router.ts:347]: Same is done across actions in insights methods

**Review by ericrommel — COMMENTED** [review]
Here are some findings. Do you think that they are blockers?

https://docs.google.com/document/d/1mdbU1_vZHksymXz63_VdQGBq-ESaNid-MAsxqUqEFMo/edit?usp=sharing

**Review by ericrommel — DISMISSED** [review]
From the feature perspective, everything looks great. I found issues related to the graphic component that can be addressed as a follow-up PR.
- https://github.com/calcom/cal.com/issues/9696
- https://github.com/calcom/cal.com/issues/9697
- https://github.com/calcom/cal.com/issues/9698

**Review by ericrommel — DISMISSED** [review]
LGTM. Just need to check the failed builds to have this PR all green.

**Review by keithwillcode — APPROVED** [review]

**Pull request merged**